### ISSUE TITLE
Using aws-vault login to GovCloud fails when assuming role on separate account

### ISSUE THREAD

**Issue by bradnussbaum** [99designs/aws-vault#374]
Running an `aws-vault login` command that assumes a role on a separate AWS account from the source profile is failing for `aws-vault` while it is working for AWS Console and CLI commands. This functionality seems to be well supported by AWS Console and CLI so I expect it's possible and just needs to be implemented or fixed in aws-vault.

I have two AWS accounts:
1) 073429735469 - to be accessed by profile `govtech`
2) 353892945261 - to be accessed by profile `govdev`

I have two profiles setup to access them. The first one is the source profile `govtech` that is configured with MFA and does in fact work with `aws-vault login`. The second is `govdev` which assumes an Administrator role on account `353892945261` using the source profile `govtech`. I am able to Switch Role in the AWS GovCloud console from account 073429735469 to account 353892945261 using this Administrator role so I know it's working.

Here are the profiles I have locally:
```
[profile govtech]
output = json
region = us-gov-west-1
mfa_serial = arn:aws-us-gov:iam::073429735469:mfa/machine

[profile govdev]
source_profile = govtech
role_arn = arn:aws-us-gov:iam::353892945261:role/Administrator
```

I am able to login to the console of account 073429735469 using the govtech profile as follows:
```
$ aws-vault login govtech --debug
2019/05/24 15:47:54 [keyring] Considering backends: [keychain pass file]
2019/05/24 15:47:54 Loading config file /Users/machine/.aws/config
2019/05/24 15:47:54 Parsing config file /Users/machine/.aws/config
2019/05/24 15:47:54 Skipping session token and using master credentials directly
2019/05/24 15:47:54 Looking up keyring for govtech
2019/05/24 15:47:54 [keyring] Querying keychain for service="aws-vault", account="govtech", keychain="aws-vault.keychain"
2019/05/24 15:47:57 [keyring] Found item "aws-vault (govtech)"
2019/05/24 15:47:57 No session token found, calling GetFederationToken
2019/05/24 15:47:58 Creating login token, expires in 12h0m0s
```

When I try to login to the account 353892945261 with the govdev profile it fails as follows:
```
$ aws-vault login govdev --debug
2019/05/24 15:40:01 [keyring] Considering backends: [keychain pass file]
2019/05/24 15:40:01 Loading config file /Users/machine/.aws/config
2019/05/24 15:40:01 Parsing config file /Users/machine/.aws/config
2019/05/24 15:40:01 Skipping session token and using master credentials directly
2019/05/24 15:40:01 Looking up keyring for govtech
2019/05/24 15:40:01 [keyring] Querying keychain for service="aws-vault", account="govtech", keychain="aws-vault.keychain"
2019/05/24 15:40:01 [keyring] Found item "aws-vault (govtech)"
2019/05/24 15:40:01 Assuming role arn:aws-us-gov:iam::353892945261:role/Administrator with iam credentials
aws-vault: error: Failed to get credentials for govtech (source profile for govdev): InvalidClientTokenId: The security token included in the request is invalid.
	status code: 403, request id: b7e4b1a2-7e5b-12e9-81b4-eb17234e2f7d
```

I am able to use the aws-cli to execute commands against the govdev profile as follows:
```
$ aws --profile govdev --region us-gov-west-1 s3 ls
// output omitted
```

Couple of other thoughts:
1. This issue is only happening for GovCloud. It's working fine for Public Cloud using region `us-west-2`.
2. This issue also happens with `exec` calls so it's not just a `login` issue

**Comment by FernandoMiguel** [comment]
Don't you need MFA on the second account?

--
Fernando Miguel

**Comment by bradnussbaum** [comment]
The MFA should only need to be specified in the source profile “govtech” which manages the authentication. That seems to be the case for public cloud cases that I have working. This config also works with aws-cli currently. If I follow the debug logs correctly aws-vault seems to be accessing the credentials for “govtech” and then failing to assume the role. This could happen for simple reasons like a request now using the region from the source profile which is set at us-gov-west-1.  Could be other reasons as well but setting region is one area that often trips things up.

**Comment by jstewmon** [comment]
AFAIK, the aws-cli (botocore) will only use the `mfa_serial` setting when it appears alongside `role_arn`. See: https://github.com/boto/botocore/blob/develop/botocore/credentials.py#L1169

Maybe the public cloud cases to which you referred do not require `aws:MultiFactorAuthPresent` to assume role?

aws-vault was recently updated use the `mfa_serial` setting when it appears on the source_profile, but a release has yet to be made with that feature, so until a new release is made, you will need to set `mfa_serial` on your assume role profile or build aws-vault from source to use `mfa_serial` on your source_profile.

Not sure if this applied to you, but when you use the `login` command, you can't login to the console as the user principal for your master credentials and pass any MFA checks because AWS does not allow console login with temporary credentials for a user principal, and the api for generating the login url does not accept an mfa token.

**Comment by bradnussbaum** [comment]
After you posted this I started testing changes to the "govdev" profile where I added in output, region and mfa_serial. I believe the intention is that these should get inherited from the source_profile but as you point out that might not be the case. In short summary, I now have `aws-vault exec` and `aws-vault login` commands working and the primary issue is actually the `region` needing to be specified on each profile. Specifying `mfa_serial` wasn't needed for me and when present caused me to enter codes every time I ran a command. Here's a short summary of what I discovered:

1. The `region` config is *not inherited* which causes a fallback to using https://sts.amazonaws.com/ instead of https://sts.us-gov-west-1.amazonaws.com.  When `region` is specified in `govdev` as below then it all works.
```
[profile govtech]
output = json
region = us-gov-west-1
mfa_serial = arn:aws-us-gov:iam::073429735469:mfa/machine

[profile govdev]
output = json
region = us-gov-west-1
source_profile = govtech
role_arn = arn:aws-us-gov:iam::353892945261:role/Administrator
```

2. The `mfa_serial` config is *not required* on profiles using a `source_profile`. When the `mfa_serial` config is present it prompts for a code to be entered on every command. When the `mfa_serial` config is absent commands may be executed without being prompted for a code each time. This is a similar behavior to the public cloud profiles.

In summary my issue is resolved and everything is working for me. There might be some work here to understand why inheritance of the `region` is happening but other than that I think this is good.

**Comment by FernandoMiguel** [comment]
Afaik, in order to use login and get a federated url you need MFA
That's why other commands work for you but not login. 

Aws-vault currently doesn't propagate mfa from the source, but there's an approved PR that will change that. 

PS it would seem your admin role doesn't enforce MFA which is a terrible security practice.

**Comment by jstewmon** [comment]
In your last paragraph, did you mean *understand why inheritance of the `region` is **not** happening*?

It sounds like you were expecting it to be inherited, but it is not, so your resolution was to add `region` to your role profile.

Neither aws-vault nor the AWS SDKs will inherit setting like region from source profiles AFIAK.

If the profile you use with aws-vault has `region`, aws-vault will provide environment variables (`AWS_REGION` and `AWS_DEFAULT_REGION`) to the subprocess, so that the subprocess will use the region from the profile.

You could add `region` (and any other non-credential settings) to a `default` profile, and the subprocess will use those as long as you do not override the profile to be used by the subprocess (i.e. with `env` or a flag).

```ini
[profile default]
region = us-gov-west-1

[profile govtech]
output = json
mfa_serial = arn:aws-us-gov:iam::073429735469:mfa/machine

[profile govdev]
source_profile = govtech
role_arn = arn:aws-us-gov:iam::353892945261:role/Administrator
```

```bash
$ aws-vault exec govdev -- aws configure get region
us-gov-west-1
```

Sounds like you use multiple regions routinely, so that may not be of much help though. 🤷‍♂

**Comment by bradnussbaum** [comment]
Thanks for the responses. I was able to learn some more things here and make some adjustments that I think I understand better now. In summary I do use multiple regions, I did realize that the `govdev` account was *not* enforcing MFA for the trusted relationship on the Administrator role to the `govtech` account. I'll try to hit each point below:

1) I did think region would be inherited from the `source_profile` initially and understand now that it is not. I also do use multiple regions routinely and my default profile is based in `us-west-2` where most of our public cloud operations reside as follows:
```
[profile default]
region = us-west-2
```

2) I was *not* properly enforcing MFA for the Administrator role on the `govdev` account. When I required MFA with the following policy then `aws-vault login` and aws-vault exec` commands failed without the `mfa_serial` config on the `govdev` role profile. This is the AWS policy I applied to `govdev` Administrator role that allows `govtech` access with MFA required now:
```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws-us-gov:iam::073429735469:root"
      },
      "Action": "sts:AssumeRole",
      "Condition": {
        "Bool": {
          "aws:MultiFactorAuthPresent": "true"
        }
      }
    }
  ]
}
```

With MFA required on `govtech` I had to add `mfa_serial` to the role profile in order to get access again. After making this change I am able to use `aws-vault exec` and `aws-vault login` calls successfully and I would think that MFA is being properly acquired and enforced now. Here's what I have in my config now:
```
[profile default]
region = us-west-2

[profile govtech]
output = json
region = us-gov-west-1
mfa_serial = arn:aws-us-gov:iam::073429735469:mfa/machine

[profile govdev]
output = json
region = us-gov-west-1
source_profile = govtech
role_arn = arn:aws-us-gov:iam::353892945261:role/Administrator
mfa_serial = arn:aws-us-gov:iam::073429735469:mfa/machine
```

I've learned some things here about profile configs I didn't realize at first. Everything seems to be working great now for me and the config is sensible. I think I just got tripped up thinking about what config was being inherited from the `source_profile`. I also initially thought the `source_profile` would be the only place I'd have to specify `mfa_serial` because that device is registered with my user in the `govtech` account and it wasn't initially obvious to me that I needed to apply that config to each profile.

I do appreciate all the help. Thanks!

**Comment by FernandoMiguel** [comment]
Well... That all might soon change
https://github.com/99designs/aws-vault/pull/369

**Comment by bradnussbaum** [comment]
Yep so looks like we will only need to specify `mfa_serial` once on the `source_profile` which I think makes sense as that's the profile that is managing the credentials and login identity.
